### PR TITLE
ENH: Add beta-binomial distribution to scipy.stats

### LINKED ISSUE #7102: Implement beta binomial distribution
The beta binomial distribution is a generalized version of the binomial in which the proportion parameter is sampled from a beta distribution.  This has wide applications in bioinformatics and computational biology where events appear to be binomial in nature but the observed variance is greater than can be inferred from a binomial model alone.  For my own use, I have written a [temporary module](https://gist.github.com/PikalaxALT/6e2a39ce689764b5c53a97c0a6aa7e18) which defines a `betabinomgen` object from a scipy discrete distribution.  However, the cdf and sf functions are imprecise since they use fixed-width bins; and the ppf is a brute force approach.  An alternative would be to invoke the appropriate hypergeometric function as suggested by Wikipedia, however this particular hypergeometric function does not appear to be implemented currently.

Related: Beta negative binomial

### PR DESCRIPTION
**Pull request by lpsinger**
#### Reference issue
Closes gh-7102.

#### What does this implement/fix?
Add the beta-binomial distribution as `scipy.stats.betabinom`.

#### Additional information

This code is adapted from the version provided by @PikalaxALT in issue #7102. The following changes were made:

* Remove the brute-force implementations of `_cdf` and `_sf`, since the base clase `rv_discrete` already provides a brute-force implementation.
* Adjust imports to match code style of `_discrete_distns.py`.
* Remove some `float()` casts that are no longer necessary in Python 3 and that would break Numpy vectorization.
* Fill in basic documentation.

### CODE REVIEW

File: scipy/stats/_discrete_distns.py
Comment (tylerjereddy) [inline scipy/stats/_discrete_distns.py]: I see this class has full line coverage by tests according to Azure, which is a simple but initial good sign.

`stats` gurus will have to check this of course..

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py]: `\binom{n}{k}` for readability

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py]: This is the only reason to import `beta` feom _continuous_distn, correct?
Then best spell out the restrictions on `a` and `b` and eliminate the import.

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py]: This is `betaln(n, k)`, correct?

File: scipy/stats/_distr_params.py
Comment (ev-br) [inline scipy/stats/_distr_params.py]: Any specific reason to not use e.g. 2.3 and 0.63 ?

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py]: Pep8 nit: no spaces around = for default parameters.

**Review by ev-br — CHANGES_REQUESTED** [review]
Looks almost ready, just a few minor comments. Might want to add some tests for limiting cases, e.g. that it becomes uniform for a=b=1

File: scipy/stats/_distr_params.py
Comment (lpsinger) [reply scipy/stats/_distr_params.py]: None at all. I copy-pasted that from another distribution.

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: No, I don't think so.

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Fixed.

File: scipy/stats/_distr_params.py
Comment (lpsinger) [reply scipy/stats/_distr_params.py]: Fixed.

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Fixed.

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Fixed.

**Comment by lpsinger** [comment]
> Might want to add some tests for limiting cases, e.g. that it becomes uniform for a=b=1

I added that test.

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [reply scipy/stats/_discrete_distns.py]: Indeed,
```
log binom(n, k) = -log(n+1) - betaln(n-k+1, k+1)
```

I suspect `betaln` can handle edge cases better (large n, large k etc)

Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Right you are. Fixed.

File: scipy/stats/_discrete_distns.py
Comment (chrisb83) [inline scipy/stats/_discrete_distns.py]: Could you add a short explanation that it is a binomial distribution with random success probability p that follows a beta distribution?

**Review by chrisb83 — COMMENTED** [review]
Thanks for creating this PR. #10884 might be helpful to see if all requirements are fulfilled. I think a tutorial file is missing.

Otherwise this looks like a nice enhancement

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Done.

**Comment by lpsinger** [comment]
> I think a tutorial file is missing.

What is a tutorial file and where does it go in GitHub?

**Review by ev-br — APPROVED** [review]

**Comment by chrisb83** [comment]
> What is a tutorial file and where does it go in the repository?

See #10386:

 - Add a tutorial doc/source/tutorial/stats/continuous_XXXXX.rst
  - Add entry to doc/source/tutorial/stats/continuous.rst.
- Update the # distributions in the example code in doc/source/tutorial/stats.rst.

**Comment by lpsinger** [comment]
> > What is a tutorial file and where does it go in the repository?
> 
> See #10386:
> 
> * Add a tutorial doc/source/tutorial/stats/continuous_XXXXX.rst
> * Add entry to doc/source/tutorial/stats/continuous.rst.
> * Update the # distributions in the example code in doc/source/tutorial/stats.rst.

Done.

**Comment by lpsinger** [comment]
The CI failures seem to be unrelated. Any other feedback?

**Comment by brandondavid** [comment]
> The CI failures seem to be unrelated. Any other feedback?

The [failed Travis CI build](https://travis-ci.org/scipy/scipy/jobs/595129963) is likely due to the issue fixed by f5f5a39.  Amending your last commit (or making a new one) should resolve this.

Judging by [the log](https://circleci.com/api/v1.1/project/github/scipy/scipy/15558/output/106/0?file=true), the CircleCI error seems due to one of the LaTeX equations in your tutorial at [discrete_betabinom.rst](https://raw.githubusercontent.com/scipy/scipy/5c6f1361244ccfb138a61883d700d03aef772c63/doc/source/tutorial/stats/discrete_betabinom.rst).  Does the last equation work without `:nowrap:`?

File: doc/source/tutorial/stats/discrete_betabinom.rst
Comment (lpsinger) [reply doc/source/tutorial/stats/discrete_betabinom.rst]: Done.

**Comment by lpsinger** [comment]
The docs build is still failing. How do I find the build.log file?

**Comment by lpsinger** [comment]
Fixed it. The build is now passing.

File: scipy/stats/_discrete_distns.py
Comment (chrisb83) [inline scipy/stats/_discrete_distns.py]: Could you add here that n >= 0, a,b >0?

File: scipy/stats/_discrete_distns.py
Comment (chrisb83) [inline scipy/stats/_discrete_distns.py]: one could even say n > 0, n=0 is a degenerate case, not sure if that could be an issue

**Review by chrisb83 — COMMENTED** [review]
PR looks good, just two minor remarks...

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: It is a degenerate case, but I tried creating an instance with `n = 0` and found that `.rvs()`, `.mean()`, `.median()`, etc. return correct values. `.ppf(0)` returns `-1`, which is nonsense.

Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: The closely related `scipy.stats.binom` distribution does permit `n = 0`, and it also has the issue that `binom(0, p).ppf(0)` returns `-1`. I think that for the sake of consistency, it's best if `betabinom` also accepts `n = 0`. If this is perceived as an actual bug, then it would be best to change the behavior of both distributions for this degenerate case in a separate pull request.

File: scipy/stats/_discrete_distns.py
Comment (lpsinger) [reply scipy/stats/_discrete_distns.py]: Fixed.

**Review by chrisb83 — APPROVED** [review]

File: scipy/stats/_discrete_distns.py
Comment (chrisb83) [reply scipy/stats/_discrete_distns.py]: Ok, this case is not really important and I am fine to leave it for consistency for now

**Comment by lpsinger** [comment]
Is this waiting for more feedback?

**Comment by rgommers** [comment]
No, looks like all that's left to do is hit the green button. PR looks good to me as well. Thanks @lpsinger, and all reviewers!

**Comment by rgommers** [comment]
@chrisb83 please feel free to merge straight away after approving a PR that's ready (or, if you just want to give others time for a final look, add a comment to that effect).

**Comment by lpsinger** [comment]
Thank you!

**Comment by yoavram** [comment]
Thank you for this enchantment.

**Pull request merged**
